workers/price-reporter/src/manager/external_executor.rs

```diff
@@ -0,0 +1,467 @@
+    cancel_channel: DefaultOption<CancelChannel>,
+}
+
+impl ExternalPriceReporterExecutor {
+    /// Creates the executor for the PriceReporter worker.
+    pub(crate) fn new(
+        job_receiver: PriceReporterReceiver,
+        config: PriceReporterConfig,
+        cancel_channel: CancelChannel,
+    ) -> Self {
+        Self {
+            subscription_states: new_async_shared(HashMap::new()),
+            config,
+            job_receiver: DefaultOption::new(Some(job_receiver)),
+            cancel_channel: DefaultOption::new(Some(cancel_channel)),
+        }
+    }
+
+    /// The execution loop for the price reporter
+    pub(crate) async fn execution_loop(mut self) -> Result<(), PriceReporterError> {
+        let mut job_receiver = self.job_receiver.take().unwrap();
+        let mut cancel_channel = self.cancel_channel.take().unwrap();
+
+        // Spawn WS handler loop, which forwards reads/writes over channels
+        let (msg_out_tx, mut msg_in_rx) = self.spawn_ws_handler_loop();
+
+        loop {
+            tokio::select! {
+                // Process price update from external price reporter
+                Some(price_message) = msg_in_rx.recv() => {
+                    self.handle_price_update(price_message).await.map_err(PriceReporterError::ExchangeConnectionError)?;
+                }
+
+                // Dequeue the next job from elsewhere in the local node
+                Some(job) = job_receiver.recv() => {
+                    if self.config.disabled {
+                        warn!("ExternalPriceReporter received job while disabled, ignoring...");
+                        continue;
+                    }
+
+                    tokio::spawn({
+                        let mut self_clone = self.clone();
+                        let msg_out_tx = msg_out_tx.clone();
+                        async move {
+                            if let Err(e) = self_clone.handle_job(job, msg_out_tx).await {
+                                error!("Error in ExternalPriceReporter execution loop: {e}");
+                            }
+                        }.instrument(info_span!("handle_job"))
+                    });
+                },
+
+                // Await cancellation by the coordinator
+                _ = cancel_channel.changed() => {
+                    info!("ExternalPriceReporter cancelled, shutting down...");
+                    return Err(PriceReporterError::Cancelled("received cancel signal".to_string()));
+                }
+            }
+        }
+    }
+
+    /// Spawns the task responsible for handling the websocket connection
+    /// with the external price reporter
+    fn spawn_ws_handler_loop(
+        &self,
+    ) -> (UnboundedSender<WebsocketMessage>, UnboundedReceiver<PriceMessage>) {
+        let price_reporter_url: Url = self
+            .config
+            .price_reporter_url
+            .clone()
+            .unwrap()
+            .parse()
+            .expect("Invalid price reporter URL");
```

Comment:
Maybe we can parse this in the config or something so that we don't introduce a panic here for a misconfigured relayer?